src/test/transaction_tests.cpp

```diff
@@ -851,24 +853,24 @@ BOOST_AUTO_TEST_CASE(test_IsStandard)
     CheckIsNotStandard(t, "scriptpubkey");
 
     // MAX_OP_RETURN_RELAY-byte TxoutType::NULL_DATA (standard)
-    t.vout[0].scriptPubKey = CScript() << OP_RETURN << ParseHex("04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef3804678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38");
+    t.vout[0].scriptPubKey = CScript() << OP_RETURN << Vec(HexLiteral<uint8_t>("04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef3804678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38"));
     BOOST_CHECK_EQUAL(MAX_OP_RETURN_RELAY, t.vout[0].scriptPubKey.size());
     CheckIsStandard(t);
 
     // MAX_OP_RETURN_RELAY+1-byte TxoutType::NULL_DATA (non-standard)
-    t.vout[0].scriptPubKey = CScript() << OP_RETURN << ParseHex("04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef3804678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef3800");
+    t.vout[0].scriptPubKey = CScript() << OP_RETURN << Vec(HexLiteral<uint8_t>("04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef3804678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef3800"));
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1720757880

> Is `Vec` needed here because the existing `ToByteVector` has a long name? Or because of some funny visual studio exception? If none, maybe we could get rid of either.

CScript only supports << with std::vector, not std::array. It would be natural for it support std::array too, but I think that would be better be done in a followup improving CScript so this PR can be focused on replacing runtime hex parsing with compile time parsing.

Reply:
I mean both Vec and ToByteVector would work here, right?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1720801552

> I mean both `Vec(HexLiteral(` and the existing `ToByteVector(HexLiteral(` would work here, right?

Right, ToByteVector could be used here instead of Vec, and Vec was just chosen because it is shorter. Sorry I misunderstood your original question.

I do think in general Vec has some benefits over ToByteVector:

- `Vec` only changes the type of the container, not the type of the elements inside the container, so it seems preferable to use in cases like this where the internal type is not changing.

- The name ToByteVector seems not great because it returns a vector of unsigned char rather than a vector of std::byte. If intention is to be switched to std::byte later though then I think it is a good name.

But nice thing is neither of these should be neccessary if cscript supports std::array

Reply:
> But nice thing is neither of these should be necessary if cscript supports std::array

Would it be too early for me to provide a separate PR for that?

Reply:
I haven't seen https://github.com/bitcoin/bitcoin/commit/5be34598c4683b2b44f607b28592a9e68e089761
It seems to me there's a simpler solution than that, since the method contains two separate concerns (inserting size + inserting elements) - which are using different parts of the vector/array.
```patch
diff --git a/src/script/script.h b/src/script/script.h
--- a/src/script/script.h	(revision d5eed066d33780b12c1f0813a2adf021eac0ca5d)
+++ b/src/script/script.h	(date 1724057506103)
@@ -412,6 +412,34 @@
 /** Serialized script, used inside transaction inputs and outputs */
 class CScript : public CScriptBase
 {
+private:
+    void InsertSize(const size_t size)
+    {
+        if (size < OP_PUSHDATA1)
+        {
+            insert(end(), static_cast<unsigned char>(size));
+        }
+        else if (size <= 0xff)
+        {
+            insert(end(), OP_PUSHDATA1);
+            insert(end(), static_cast<unsigned char>(size));
+        }
+        else if (size <= 0xffff)
+        {
+            insert(end(), OP_PUSHDATA2);
+            uint8_t _data[2];
+            WriteLE16(_data, size);
+            insert(end(), _data, _data + sizeof(_data));
+        }
+        else
+        {
+            insert(end(), OP_PUSHDATA4);
+            uint8_t _data[4];
+            WriteLE32(_data, size);
+            insert(end(), _data, _data + sizeof(_data));
+        }
+    }
+
 protected:
     CScript& push_int64(int64_t n)
     {
@@ -465,29 +493,15 @@
 
     CScript& operator<<(const std::vector<unsigned char>& b) LIFETIMEBOUND
     {
-        if (b.size() < OP_PUSHDATA1)
-        {
-            insert(end(), (unsigned char)b.size());
-        }
-        else if (b.size() <= 0xff)
-        {
-            insert(end(), OP_PUSHDATA1);
-            insert(end(), (unsigned char)b.size());
-        }
-        else if (b.size() <= 0xffff)
-        {
-            insert(end(), OP_PUSHDATA2);
-            uint8_t _data[2];
-            WriteLE16(_data, b.size());
-            insert(end(), _data, _data + sizeof(_data));
-        }
-        else
-        {
-            insert(end(), OP_PUSHDATA4);
-            uint8_t _data[4];
-            WriteLE32(_data, b.size());
-            insert(end(), _data, _data + sizeof(_data));
-        }
+        InsertSize(b.size());
+        insert(end(), b.begin(), b.end());
+        return *this;
+    }
+
+    template<size_t N>
+    CScript& operator<<(const std::array<const unsigned char, N>& b) LIFETIMEBOUND
+    {
+        InsertSize(N);
         insert(end(), b.begin(), b.end());
         return *this;
     }
```

i.e. for `vector`:
```C++
InsertSize(b.size());
insert(end(), b.begin(), b.end());
```
and for `array`:
```C++
InsertSize(N);
insert(end(), b.begin(), b.end());
```

> that this should be a follow-up PR

I'm fine with both, I'm just providing alternatives.

Reply:
I like that my version just forwards everything in one statement, while your leaves 3, duplicated. But either one works for me.

Reply:
In my suggestion there is not `std::span` involved and the size + data writing parts are split.

Reply:
I'm fine with any of those - having a slight preference for the one which clarifies the two separate roles of the serialization - length + values

Reply:
Split this out to a separate PR: [`176d42b` (#30765)](https://github.com/bitcoin/bitcoin/pull/30765/commits/176d42b9308b439e1acd6e710a3f7bee01c56069)